tests/gocase/unit/pubsub/pubsub_test.go

```
@@ -0,0 +1,356 @@
+
+		msg := receiveType(t, pubsub, &redis.Message{})
+		require.Equal(t, "chan1", msg.Channel)
+		require.Equal(t, "hello", msg.Payload)
+		msg = receiveType(t, pubsub, &redis.Message{})
+		require.Equal(t, "chan2", msg.Channel)
+		require.Equal(t, "world", msg.Payload)
+
+		// unsubscribe from one of the channels
+		require.NoError(t, pubsub.Unsubscribe(ctx, "chan1"))
+		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
+		require.EqualValues(t, 1, rdb.Publish(ctx, "chan2", "world").Val())
+
+		// unsubscribe from the remaining channel
+		require.NoError(t, pubsub.Unsubscribe(ctx, "chan2"))
+		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
+		require.EqualValues(t, 0, rdb.Publish(ctx, "chan2", "world").Val())
+	})
+
+	t.Run("PUBLISH/SUBSCRIBE with two clients", func(t *testing.T) {
+		c1 := srv.NewClient()
+		defer func() { require.NoError(t, c1.Close()) }()
+		c2 := srv.NewClient()
+		defer func() { require.NoError(t, c2.Close()) }()
+
+		p1 := c1.Subscribe(ctx, "chan1")
+		p2 := c2.Subscribe(ctx, "chan1")
+		require.EqualValues(t, 1, receiveType(t, p1, &redis.Subscription{}).Count)
+		require.EqualValues(t, 1, receiveType(t, p2, &redis.Subscription{}).Count)
+
+		require.EqualValues(t, 2, rdb.Publish(ctx, "chan1", "hello").Val())
+		msg := receiveType(t, p1, &redis.Message{})
+		require.Equal(t, "chan1", msg.Channel)
+		require.Equal(t, "hello", msg.Payload)
+		msg = receiveType(t, p2, &redis.Message{})
+		require.Equal(t, "chan1", msg.Channel)
+		require.Equal(t, "hello", msg.Payload)
+	})
+
+	t.Run("PUBLISH/SUBSCRIBE after UNSUBSCRIBE without arguments", func(t *testing.T) {
+		pubsub := rdb.Subscribe(ctx, "chan1", "chan2", "chan3")
+		require.EqualValues(t, 1, receiveType(t, pubsub, &redis.Subscription{}).Count)
+		require.EqualValues(t, 2, receiveType(t, pubsub, &redis.Subscription{}).Count)
+		require.EqualValues(t, 3, receiveType(t, pubsub, &redis.Subscription{}).Count)
+
+		require.NoError(t, pubsub.Unsubscribe(ctx))
```

@tisonkun looks like you need to add channels here - "chan1", "chan2", "chan3". Not sure why the test is not failing earlier though and you still receive `unsubscribe` events... Perhaps some magic by kvrocks? I wonder if Redis behaves the same.

---

@vmihailenco from [Redis command docs](https://redis.io/commands/unsubscribe):

> When no channels are specified, the client is unsubscribed from all the previously subscribed channels. In this case, a message for every unsubscribed channel will be sent to the client.

... from go-redis method docs:

```go
// Unsubscribe the client from the given channels, or from all of
// them if none is given.
func (c *PubSub) Unsubscribe(ctx context.Context, channels ...string) error {
```

So I think this command should behave as "unsubscribe all" otherwise it may be a bug.

The interesting point is that, even we check the unsubscription event happened, we can still publish to those channels. @git-hulk this can be a multi-thread concurrency issue for Kvrocks specifically - I'm unsure about it.

---

>from go-redis method docs:

Indeed! I honestly can't find where that is implemented and can't check it right away so don't trust it blindly :)

Perhaps it indeed unsubscribes thanks to Redis/kvrocks behavior, but does not reset the internal state of `PubSub` which can be a cause for surprises.

---

Interesting. I can reproduce it locally and when I rename the channel name from "chan2" to "xx" the error vanished.

```diff
diff --git a/tests/gocase/unit/pubsub/pubsub_test.go b/tests/gocase/unit/pubsub/pubsub_test.go
index b48b10a..97b807b 100644
--- a/tests/gocase/unit/pubsub/pubsub_test.go
+++ b/tests/gocase/unit/pubsub/pubsub_test.go
@@ -73,11 +73,26 @@ func TestPubSub(t *testing.T) {
 
 		// unsubscribe from one of the channels
 		require.NoError(t, pubsub.Unsubscribe(ctx, "chan1"))
+		require.EqualValues(t, &redis.Subscription{
+			Kind:    "unsubscribe",
+			Channel: "chan1",
+			Count:   1,
+		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
 		require.EqualValues(t, 1, rdb.Publish(ctx, "chan2", "world").Val())
 
+		// drain incoming messages
+		// TODO: we should receive once here, but this bug cause the below unstable test steady to fail.
+		receiveType(t, pubsub, &redis.Message{})
+		receiveType(t, pubsub, &redis.Message{})
+
 		// unsubscribe from the remaining channel
 		require.NoError(t, pubsub.Unsubscribe(ctx, "chan2"))
+		require.EqualValues(t, &redis.Subscription{
+			Kind:    "unsubscribe",
+			Channel: "chan2",
+			Count:   0,
+		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan2", "world").Val())
 	})
@@ -103,7 +118,7 @@ func TestPubSub(t *testing.T) {
 	})
 
 	t.Run("PUBLISH/SUBSCRIBE after UNSUBSCRIBE without arguments", func(t *testing.T) {
-		pubsub := rdb.Subscribe(ctx, "chan1", "chan2", "chan3")
+		pubsub := rdb.Subscribe(ctx, "chan1", "xx", "chan3")
 		require.EqualValues(t, 1, receiveType(t, pubsub, &redis.Subscription{}).Count)
 		require.EqualValues(t, 2, receiveType(t, pubsub, &redis.Subscription{}).Count)
 		require.EqualValues(t, 3, receiveType(t, pubsub, &redis.Subscription{}).Count)
@@ -116,7 +131,7 @@ func TestPubSub(t *testing.T) {
 		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, &redis.Subscription{
 			Kind:    "unsubscribe",
-			Channel: "chan2",
+			Channel: "xx",
 			Count:   1,
 		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, &redis.Subscription{
@@ -126,7 +141,7 @@ func TestPubSub(t *testing.T) {
 		}, receiveType(t, pubsub, &redis.Subscription{}))
 
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
-		require.EqualValues(t, 0, rdb.Publish(ctx, "chan2", "hello").Val())
+		require.EqualValues(t, 0, rdb.Publish(ctx, "xx", "hello").Val())
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan3", "hello").Val())
 	})
```

---

Forgive my comment above, it's about stateful tests.

---

You can add `defer pubsub.Close()` to the tests to properly cleanup pubsub and go-redis should fix https://github.com/go-redis/redis/issues/2248

---

So now I get the point:

Because we don't wait for unsubscription in L76, the "chan2" subscription will never be unsubscribed and leak.

---

We may try to fail fast or reduce share state...But pubsub state are on the server side so the only approach is start a new server for each case?

---

@vmihailenco thank you! Let me think of it.